Ishai Marom - Kafka Exercise Section 4.2

Ishai-Marom opened the pull request:
Section 4.2, contains the full exercise and fixes from comments of
section 4.1.

Ishai-Marom committed:
Ishai Marom - Kafka Exercise Section 4.2

Section 4.2, contains the full exercise and fixes from comments of
section 4.1.

Ishai-Marom committed:
Ishai Marom - Kafka Exercise section 4.2

section 4.2, fixed according to the reviews on section 4.1

Ishai-Marom committed:
Ishai Marom - Kafka Exercise Section 4.2

section 4.2, fixed according to the reviews on section 4.1.
recen change: the names of the Consumer and Producer classes.

asafbena commented on Exercises/KafkaExercise/Exercise_4_2/AvroClassesGenerator/src/main/avro/person.avsc:
```diff
@@ -0,0 +1,33 @@
+      {
+        "name": "firstName",
+        "type": "string",
+        "default": "default"
```
You do not need to provide default value for mandatory fields

asafbena commented on Exercises/KafkaExercise/Exercise_4_2/AvroClassesGenerator/src/main/avro/person.avsc:
```diff
@@ -0,0 +1,33 @@
+      {
+        "name": "lastName",
+        "type": "string",
+        "default": "default"
```
Same

asafbena commented on Exercises/KafkaExercise/Exercise_4_2/AvroClassesGenerator/src/main/avro/person.avsc:
```diff
@@ -0,0 +1,33 @@
+      {
+        "name": "countryOfBirth",
+        "type": "string",
+        "default": "default"
```
The default value required to be Israel

asafbena commented on Exercises/KafkaExercise/Exercise_4_2/AvroClassesGenerator/src/main/avro/person.avsc:
```diff
@@ -0,0 +1,33 @@
+      {
+        "name": "age",
+        "type": "int",
+        "default": -1
```
No default value

asafbena commented on Exercises/KafkaExercise/Exercise_4_2/AvroClassesGenerator/src/main/avro/person.avsc:
```diff
@@ -0,0 +1,33 @@
+      {
+        "name": "petName",
+        "type": "string",
+        "default": "default"
```
Look again about the requirement of this field

asafbena commented on Exercises/KafkaExercise/Exercise_4_2/Consumer/src/main/java/org/ishaym/training/KafkaEnvironmentSetUp.java:
```diff
@@ -0,0 +1,78 @@
+import java.util.concurrent.TimeoutException;
+
+
+public class KafkaEnvironmentSetUp {
```
Create AdminClient instance and use it across the class

asafbena commented on Exercises/KafkaExercise/Exercise_4_2/Consumer/src/main/java/org/ishaym/training/MessagesConsumer.java:
```diff
@@ -0,0 +1,51 @@
+        props.put("bootstrap.servers", constants.getKafkaProperties().getBootstrapServer());
+        props.put("schema.registry.url", constants.getKafkaProperties().getSchemaRegistryUrl());
+        props.put("group.id", constants.getConsumerProperties().getGroupId());
+        props.put("key.deserializer", "org.apache.kafka.common.serialization.IntegerDeserializer");
```
Extract to constants file

asafbena commented on Exercises/KafkaExercise/Exercise_4_2/Consumer/src/main/java/org/ishaym/training/runnable/ConsumerAction.java:
```diff
@@ -0,0 +1,45 @@
+                    String output = MessageFormat.format("message key: {0} , message value: {1}",
+                            record.key(), record.value());
+                    LOGGER.info(output);
+                }
```
Extract to function

asafbena commented on Exercises/KafkaExercise/Exercise_4_2/Producer/src/main/java/org/ishaym/training/KafkaEnvironmentSetUp.java:
```diff
@@ -0,0 +1,53 @@
+            throws ExecutionException, InterruptedException, IOException {
+        LOGGER.debug("checking if topic already exists");
+
+        try (AdminClient adminClient = AdminClient.create(createAdminProperties())) {
```
You should run it for a specific amount of times in order to support a case in which the producer started before the consumer

asafbena commented on Exercises/KafkaExercise/Exercise_4_2/Producer/src/main/java/org/ishaym/training/KafkaEnvironmentSetUp.java:
```diff
@@ -0,0 +1,53 @@
+public class KafkaEnvironmentSetUp {
+    private static final Logger LOGGER = LogManager.getLogger(KafkaEnvironmentSetUp.class);
+
+    private KafkaEnvironmentSetUp() {
```
Same as above

Ishai-Marom committed:
Ishai Marom - Kafka Exercise section 4.2

section 4.2, fixed according to review, and changed the config classes
according to plugin that creates jackson classes automaticly from json file.

asafbena commented on Exercises/KafkaExercise/Exercise_4_2/Consumer/src/main/java/org/ishaym/training/KafkaEnvironmentSetUp.java:
```diff
@@ -0,0 +1,83 @@
+            LOGGER.fatal(e);
+            System.exit(-1);
+        }
+    }
```
Use singleton way or just create it inside the constructor which is the favourite way in this case

Ishai-Marom committed:
Ishai Marom - Kafka Exercise section 4.2

    section 4.2, fixed according to review, and changed the config classes
    according to plugin that creates jackson classes automaticly from json file.
recent change: formating files.

asafbena commented on Exercises/KafkaExercise/Exercise_4_2/Consumer/src/main/java/org/ishaym/training/MessagesConsumer.java:
```diff
@@ -0,0 +1,54 @@
+    }
+
+    public void consume() {
+        Thread thread = new Thread(new ConsumerAction(consumer));
```
The thread instance should be exposed to outside of the function so it will be accessible in a case we will want to operate it in some way

asafbena commented on Exercises/KafkaExercise/Exercise_4_2/Consumer/src/main/java/org/ishaym/training/runnable/ConsumerAction.java:
```diff
@@ -0,0 +1,47 @@
+        this.consumer = consumer;
+    }
+
+    private void action(ConsumerRecords<Integer, Person> records) {
```
You need to change to a name which describes in a better way what the function is responsible for

asafbena commented on Exercises/KafkaExercise/Exercise_4_2/Consumer/src/main/java/org/ishaym/training/runnable/ConsumerAction.java:
```diff
@@ -0,0 +1,47 @@
+import java.text.MessageFormat;
+import java.time.Duration;
+
+public class ConsumerAction implements Runnable {
```
There is no need for the word 'Action', Just call the class 'KafkaConsumer'

asafbena commented on Exercises/KafkaExercise/Exercise_4_2/Producer/src/main/java/org/ishaym/training/KafkaEnvironmentSetUp.java:
```diff
@@ -0,0 +1,80 @@
+            LOGGER.fatal(e);
+            System.exit(-1);
+        }
+    }
```
Same as above

asafbena commented on Exercises/KafkaExercise/Exercise_4_2/Producer/src/main/java/org/ishaym/training/Main.java:
```diff
@@ -0,0 +1,34 @@
+            MessagesProducer producer = new MessagesProducer();
+            producer.sendMessage(
+                    Constants.genInstance().getConfigurations().getTopicProperties().getName(), 0,
+                    new Person("first name", "last name", "Israel", 10, null));
```
Extract to constants

Ishai-Marom committed:
Ishai Marom - Kafka Exercise section 4.2

section 4.2, fixed according to review.

Ishai-Marom committed:
Ishai Marom - Kafka Exercise section 4.2

section 4.2, fixed according to review on section 4.3.

Ishai-Marom committed:
Ishai Marom - Kafka Exercise section 4.2

section 4.2, fixed according to review on section 4.3.
